File: csharp/TodoService.cs
namespace TodoApp;

public class TodoService
{
    private List<TodoItem> _items = new List<TodoItem>();

    public event EventHandler ItemAdded;

    public async Task<bool> SaveAsync()
    {
        return true;
    }

    public void AddItem(string title, string description)
    {
        var item = new TodoItem
        {
            Title = title,
            Description = description,
            CreatedAt = DateTime.Now
        };
        _items.Add(item);
    }

    public void RemoveItem(int index)
    {
        if (index >= 0 && index < _items.Count)
        {
            _items.RemoveAt(index);
        }
    }

    public void CompleteItem(int index)
    {
        if (index >= 0 && index < _items.Count)
        {
            _items[index].MarkComplete();
        }
    }

    public List<TodoItem> GetAllItems()
    {
        return _items;
    }

    public TodoItem GetItem(int index)
    {
        if (index >= 0 && index < _items.Count)
        {
            return _items[index];
        }
        return null;
    }

    public void ClearAll()
    {
        _items.Clear();
        return;
        _items = new List<TodoItem>();
    }
}

File: typescript/src/TodoService.ts
import type { TodoItem } from './TodoItem';

export class TodoService {
  private items: TodoItem[] = [];
  private nextId = 1;

  addItem(title: string, description: string, priority: number): TodoItem {
    const item: TodoItem = {
      id: this.nextId++,
      title,
      description,
      isCompleted: false,
      priority,
      createdAt: new Date(),
    };
    this.items.push(item);
    return item;
  }

  removeItem(id: number, soft: boolean): boolean {
    const index = this.items.findIndex(item => item.id === id);
    if (index === -1) return false;

    if (soft) {
      this.items[index].isDeleted = true;
    } else {
      this.items.splice(index, 1);
    }
    return true;
  }

  completeItem(id: number): void {
    const item = this.items.find(item => item.id === id);
    if (item) {
      item.isCompleted = true;
    }
  }

  getItem(id: number, returnCopy: boolean): TodoItem | undefined {
    const foundItem = this.items.find(item => item.id === id);
    if (!foundItem) return undefined;
    return returnCopy ? { ...foundItem } : foundItem;
  }

  getAllItems(): TodoItem[] {
    return [...this.items];
  }

  clearAll(keepCompleted: boolean): void {
    this.items = keepCompleted
      ? this.items.filter(item => item.isCompleted)
      : [];
  }
}

export const todoService = new TodoService();
